pkg/mailboxwatch/interface.go

```diff
@@ -0,0 +1,77 @@
+
+	"k8s.io/apimachinery/pkg/runtime"
+	restclient "k8s.io/client-go/rest"
+	upstreamcache "k8s.io/client-go/tools/cache"
+
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	kcpinformers "github.com/kcp-dev/apimachinery/v2/third_party/informers"
+	kcpscopedclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpapiinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	tenancyv1a1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+)
+
+// NewSharedInformer constructs a shared index informer on objects of a given kind in mailbox workspaces.
+// It follows the usual pattern for a constructor of informers except for an additional parameter
+// at the front, which it uses to keep appraised of the mailbox workspaces.
+// The ListerWatcher can be constructed using `upstreamcache.NewListWatchFromClient`.
+func NewSharedInformer(
+	mailboxWorkspacePreInformer tenancyv1a1informers.WorkspaceInformer,
+	listerWatcher upstreamcache.ListerWatcher,
+	exampleObject runtime.Object,
+	defaultEventHandlerResyncPeriod time.Duration,
+	indexers upstreamcache.Indexers,
+) kcpcache.ScopeableSharedIndexInformer {
+	if true {
+		panic("not implemented yet")
+	}
+	// Implementation outline:
+	// Use the informer on mailbox workspaces to stay appraised of the logicalcluster.Name for
+	// each mailbox workspace qua logical cluster.
+	// Wrap the given ListerWatcher with filtering that passes only objects in one of those logical clusters.
+	// At first stage of development, do not take special care when mailbox workspaces arrive or depart,
+	// because objects in them will likely arrive later and depart sooner.
+	// At a later stage of development, do not rely on happy timing.
+	var wrappedListerWatcher upstreamcache.ListerWatcher
+	return kcpinformers.NewSharedIndexInformer(wrappedListerWatcher, exampleObject, defaultEventHandlerResyncPeriod, indexers)
+}
+
+// NewSharedInformerForEdgeConfig is like NewSharedInformer but takes a REST Config for the
+// edge service provider workspace and constructs the pre-informer for the mailbox workspaces.
+// That pre-informer is then used to call NewSharedInformer and is also returned,
+// so that the caller can wait on HasSynced of the mailbox workspace informer.
+func NewSharedInformerForEdgeConfig(
+	edgeServiceProviderWorkspaceClientConfig *restclient.Config,
+	listerWatcher upstreamcache.ListerWatcher,
+	exampleObject runtime.Object,
+	defaultEventHandlerResyncPeriod time.Duration,
+	indexers upstreamcache.Indexers,
+) (tenancyv1a1informers.WorkspaceInformer, kcpcache.ScopeableSharedIndexInformer, error) {
+	workspaceScopedClientset, err := kcpscopedclientset.NewForConfig(edgeServiceProviderWorkspaceClientConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create clientset for mailbox workspaces: %w", err)
+	}
+	workspaceScopedInformerFactory := kcpapiinformers.NewSharedScopedInformerFactoryWithOptions(workspaceScopedClientset, 0)
```

Comment:
Do we need to allow setting Rsync period also for the pre- informer ?

Reply:
The default resync is on the factory, not the pre-informer.
As far as I know, resync is an obsolete feature.  I do not understand how a correct controller would benefit from using resync.  At best it could paper over _some_ flaws in an incorrect controller.  Resync does not notify of deletions.

Would the C2P Controller use resync?